### PR TITLE
[Core][Flink][Spark]: Refactor `TaskWriter` implementations

### PR DESCRIPTION
**Pull request by dungdm93**
This PR aim to:
1. replace `BaseTaskWriter` inner classes (`BaseRollingWriter`, `RollingFileWriter`, `RollingEqDeleteWriter`)
by the respective implementations of `RollingFileWriter` interface.
2. provide single implementation of `TaskWriter` that can handle both partition & un-partition data (by delegate to the `PartitioningWriter`)

Here is my approach, from top down:
```txt
TaskWriter
    |
    V
PartitioningWriter
    |
    V
RollingFileWriter
    |
    V
FileWriter
    |
    V
FileAppender
```

1. `TaskWriter` is used to handle diffent kinds of record.
    * With Insert-only data, `TaskWriter` basically call `PartitioningWriter.write`. See [`DirectTaskWriter`](https://github.com/apache/iceberg/blob/18bc27070661c71d5590b3bed39b92a51d6b4c2d/core/src/main/java/org/apache/iceberg/io/DirectTaskWriter.java) for more details.
    * With delta data, `TaskWriter` can have 3 `PartitioningWriter`s for `insertWriter`, `equalityDeleteWriter` and `positionDeleteWriter`. So, for each incomming record, base on its type (insert, update or delete), TaskWriter will call corresponding writer to write data. See [`FlinkTaskWriter`](https://github.com/apache/iceberg/blob/18bc27070661c71d5590b3bed39b92a51d6b4c2d/flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkTaskWriter.java) for more details.
2. `PartitioningWriter` is used to write to multiple specs and partitions.
Note that for unpartitioned tables, `partition = null` is passed to `PartitioningWriter.write`
3. Internally, `PartitioningWriter` already use `RollingFileWriter` for rolling to new file if current file is large. `RollingFileWriter` is a just wrapper of other `FileWriter`.
4. `FileWriter` is used to write to single file.

# Why we need that
* The scope that inner classes of bounded to outer object (`BaseTaskWriter` instance) that make the code more complex and hard to read and understand.
* Currently, the approach of separated class for partition & un-partition data is still OK. But it's better if we can unify it into a single class, this mean:
    * Less effort is needed for new engine
    * No need to create new class when existing engine wanna extends new use-case. For example, now Flink only support fan-out partition which is required for streaming execution mode, but when execution mode is batch, people may prefer using `ClusteredPartitionWriter` because it take less resources.

# Result
With new 2 TaskWriters `DirectTaskWriter` and `FlinkTaskWriter`, it can cover all Flink and Spark cases.
Bellow is equal code of Flink's DeltaTaskWriter:

* write INSERT only data, unpartition table:
```java
// current
taskWriter = new UnpartitionedWriter<>(...);

// new
partitioner = DirectTaskWriter.unpartition();
taskWriter = new DirectTaskWriter<>(partitioner, ...);
```

* write INSERT only data, partition table:
```java
// current
class RowDataPartitionedFanoutWriter extends PartitionedFanoutWriter<RowData> {...}
taskWriter = new RowDataPartitionedFanoutWriter(...);

// new
partitioner = FlinkTaskWriter.partitionerFor(spec, schema, flinkSchema);
taskWriter = new DirectTaskWriter<>(partitioner, ...);
```

* write both INSERT and DELETE data, unpartition table:
```java
// current
taskWriter = new UnpartitionedDeltaWriter(...);

// new
partitioner = DirectTaskWriter.unpartition();
taskWriter = new FlinkTaskWriter(partitioner, ...);
```

* write both INSERT and DELETE data, partition table:
```java
// current
taskWriter = new PartitionedDeltaWriter(...);

// new
partitioner = FlinkTaskWriter.partitionerFor(spec, schema, flinkSchema);
taskWriter = new FlinkTaskWriter(partitioner, ...);
```

# How do I test it
Pass all unit-tests and run with some sample datasets in my local machine.

### CODE REVIEW

**Comment by dungdm93** [comment]
cc @aokolnychyi, @rdblue, @jackye1995, @openinx, @stevenzwu, @szehon-ho, @RussellSpitzer

File: core/src/main/java/org/apache/iceberg/io/PartitioningWriter.java [hunk 1]
```diff
@@ -46,7 +46,7 @@
    * @param spec a partition spec
    * @param partition a partition or null if the spec is unpartitioned
    */
-  void write(T row, PartitionSpec spec, StructLike partition);
+  PathOffset write(T row, PartitionSpec spec, StructLike partition);
```
Comment (dungdm93) [inline core/src/main/java/org/apache/iceberg/io/PartitioningWriter.java:49]: Also required

**Comment by rdblue** [comment]
@dungdm93, I don't think I understand quite what you're trying to do in this PR from the description. Can you add some more detail about what your motivation is and what you're changing? It would probably also help to do this in several smaller PRs.

**Comment by dungdm93** [comment]
@rdblue sorry for my bad wording. Let's me try to add more details

**Comment by dungdm93** [comment]
Hello @rdblue, I just update the PR description, hope it's clear enough to understand.

File: api/src/main/java/org/apache/iceberg/PartitionKey.java [hunk 2]
```diff
@@ -28,11 +28,10 @@
 
 /**
  * A struct of partition values.
- * <p>
+ * <br>
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/PartitionKey.java:31]: These should be `<p>`. Could you revert this change?

File: api/src/main/java/org/apache/iceberg/PartitionKey.java [hunk 3]
```diff
@@ -28,11 +28,10 @@
...
  * Instances of this class can produce partition values from a data row passed to {@link #partition(StructLike)}.
  */
 public class PartitionKey implements StructLike, Serializable {
-
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/PartitionKey.java:35]: Can you revert unnecessary whitespace changes?

File: core/src/main/java/org/apache/iceberg/deletes/EqualityDeleteWriter.java [hunk 4]
```diff
@@ -58,8 +60,9 @@ public EqualityDeleteWriter(FileAppender<T> appender, FileFormat format, String
   }
 
   @Override
-  public void write(T row) {
+  public PathOffset write(T row) {
     appender.add(row);
+    return PathOffset.of(location, recordCount++);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/deletes/EqualityDeleteWriter.java:65]: In Iceberg, we don't use the return value of `++` operators because it is hard to read code that uses them. Can you move the increment to a separate line?

**Comment by RussellSpitzer** [comment]
The description makes sense to me here, It will take me some time though to get through this whole PR, I'll try to set aside time later this week.

File: api/src/main/java/org/apache/iceberg/PartitionKey.java (hunk 2 above)
Comment (dungdm93) [reply api/src/main/java/org/apache/iceberg/PartitionKey.java:31]: Yeah. I see, I thought it's HTML tag.

File: api/src/main/java/org/apache/iceberg/PartitionKey.java (hunk 3 above)
Comment (dungdm93) [reply api/src/main/java/org/apache/iceberg/PartitionKey.java]: Done

File: api/src/main/java/org/apache/iceberg/PartitionKey.java [hunk 5]
```diff
@@ -98,11 +98,12 @@ public String toPath() {
    * @param row a StructLike row
    */
   @SuppressWarnings("unchecked")
-  public void partition(StructLike row) {
+  public PartitionKey partition(StructLike row) {
```
Comment (openinx) [inline api/src/main/java/org/apache/iceberg/PartitionKey.java:101]: I'm think this will produce an api compatibility issue,  why do we need to change this basic API ?

File: core/src/main/java/org/apache/iceberg/io/FileWriter.java [hunk 6]
```diff
@@ -51,8 +51,9 @@ default void write(Iterable<T> rows) {
    * Writes a row to a predefined spec/partition.
    *
    * @param row a data or delete record
+   * @return PathOffset of written row
    */
-  void write(T row);
+  PathOffset write(T row);
```
[the remark this answers is not in the document]
Comment (openinx) [reply core/src/main/java/org/apache/iceberg/io/FileWriter.java:56]: You mean each abstracted FileWriter will get a `PathOffset` back when append a newly row ? That does not make sense for me because not every writer need this PathOffset to do the following thing.

File: api/src/main/java/org/apache/iceberg/PartitionKey.java (hunk 5 above)
Comment (dungdm93) [reply api/src/main/java/org/apache/iceberg/PartitionKey.java:101]: This is just sidecar change, I made it to align with other `StructLike` wrapper, `StructProjection.wrap`, `IndexedStructLike.wrap`, `InternalRowWrapper.wrap`,... to name a few.
@openinx Could you tell me why this can make an api compatibility issue?

File: core/src/main/java/org/apache/iceberg/io/FileWriter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dungdm93) [reply core/src/main/java/org/apache/iceberg/io/FileWriter.java:56]: I agree that not every writer need `PathOffset`, but it's required for writing delta data.
For the writer that does not need this, just ignore the return.

File: core/src/main/java/org/apache/iceberg/deletes/EqualityDeleteWriter.java (hunk 4 above)
Comment (dungdm93) [reply core/src/main/java/org/apache/iceberg/deletes/EqualityDeleteWriter.java]: Changed to:
```java
    long offset = recordCount++;
    return PathOffset.of(location, offset);
```

File: api/src/main/java/org/apache/iceberg/PartitionKey.java (hunk 5 above)
Comment (dungdm93) [reply api/src/main/java/org/apache/iceberg/PartitionKey.java:101]: Yeah. Let's me roll it back.
@openinx could you help me review other changes

Comment (openinx) [reply api/src/main/java/org/apache/iceberg/PartitionKey.java:101]: Yes, I'm currently checking the whole write path.  I think I will need one or two day to understand the whole newly introduced writers since 0.13.x.  Replacing the old writer API with the new one is a great thing. I think we can collaborate to make this forward. Thanks

File: core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+ */
+
+package org.apache.iceberg.io;
+
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.deletes.PositionDelete;
+import org.apache.iceberg.io.DefaultPartitioningWriterFactory.Type;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public interface PartitioningWriterFactory<T> {
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java:27]: Why name it as `PartitioningWriterFactory` ?  I don't see any partition info for those defined interface methods.

File: core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    private long targetFileSizeInBytes = 0;
+
+    public Builder(FileWriterFactory<T> writerFactory) {
+      this.writerFactory = writerFactory;
+    }
+
+    private void checkArguments() {
+      Preconditions.checkArgument(writerFactory != null, "writerFactory is required non-null");
+      Preconditions.checkArgument(fileFactory != null, "fileFactory is required non-null");
+      Preconditions.checkArgument(io != null, "io is required non-null");
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java:52]: Nit:  The `fileFormat` is also required to be not null, right ?

File: core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      return new DefaultPartitioningWriterFactory<>(writerFactory, fileFactory, io, fileFormat,
+          targetFileSizeInBytes, Type.FANOUT);
+    }
+
+    public Builder<T> fileFactory(OutputFileFactory newFileFactory) {
+      this.fileFactory = newFileFactory;
+      return this;
+    }
+
+    public Builder<T> io(FileIO newIO) {
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java:72]: It's more clear to name this builder method as `fileIO` .

File: core/src/main/java/org/apache/iceberg/io/PathOffset.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    PathOffset that = (PathOffset) o;
+    return this.offset == that.offset && this.path.equals(that.path);
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/io/PathOffset.java:74]: Please use the `Comparators.charSequences().compare(a, b)` to compare two `CharSequence` because of this [comment](https://github.com/apache/iceberg/blob/90225d6c9413016d611e2ce5eff37db1bc1b4fc5/api/src/main/java/org/apache/iceberg/types/Comparators.java#L309-L316).

There is a similar code path you can take a look. https://github.com/apache/iceberg/blob/90225d6c9413016d611e2ce5eff37db1bc1b4fc5/api/src/main/java/org/apache/iceberg/util/CharSequenceWrapper.java#L58

File: core/src/main/java/org/apache/iceberg/io/PathOffset.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("path", path)
+        .add("offset", offset)
+        .toString();
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(path, offset);
```
Comment (openinx) [inline core/src/main/java/org/apache/iceberg/io/PathOffset.java:61]: Please use `JavaHashes.hashCode(charseq)` to get the `CharSequence`'s hashCode.

File: flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkTaskWriter.java [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+    partition = partitioner.apply(row);
+    equalityKey = equalityKeyer.apply(row);
+
+    switch (row.getRowKind()) {
+      case INSERT:
+        if (upsert) {
+          delete(row);
+        }
+        insert(row);
+        break;
```
Comment (openinx) [inline flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkTaskWriter.java:129]: Nit: Please leave a empty new line between two cases code block.

File: flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkTaskWriter.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+  protected void insert(RowData row) throws IOException {
+    StructLike copiedKey = StructCopy.copy(equalityKey);
+
+    internalPosDelete(copiedKey);
+    PathOffset pathOffset = insertWriter.write(row, spec, partition);
+    insertedRowMap.put(copiedKey, pathOffset);
+  }
+
+  protected void delete(RowData row) throws IOException {
+    StructLike copiedKey = StructCopy.copy(equalityKey);
```
Comment (openinx) [inline flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkTaskWriter.java:168]: In this `delete` method, we don't need to deep copy a new `equalityKey` because we won't hold any reference in the global `insertedRowMap` or other shared memory.

File: core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java (hunk 7 above)
Comment (dungdm93) [reply core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java:27]: It's factory class used to create `PartitioningWriter`s

File: core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java (hunk 8 above)
Comment (dungdm93) [reply core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java:52]: Actually, `fileFormat` is used in `PartitioningWriter` because ORC is not support rolling yet.
But, yes. the `fileFormat` is also required to be not null

File: core/src/main/java/org/apache/iceberg/io/DirectTaskWriter.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+import java.util.function.Function;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.util.Tasks;
+
+public class DirectTaskWriter<T> implements TaskWriter<T> {
```
Comment (dungdm93) [inline core/src/main/java/org/apache/iceberg/io/DirectTaskWriter.java:28]: @openinx are you have any naming suggestion for this class, `DirectTaskWriter`, `AppendTaskWriter`,...?

File: core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java (hunk 7 above)
Comment (dungdm93) [reply core/src/main/java/org/apache/iceberg/io/PartitioningWriterFactory.java:30]: I just add a little docs for better understanding.
